pypdf/_writer.py

```diff
@@ -2110,22 +2115,76 @@ def remove_annotations(
+                        annotation: ArrayObject,
+                        obj: DictionaryObject
+                    ) -> bool:
+                        try:
+                            uri = obj['/A']['/URI']
+                            return uri.startswith('https://google.com/')
+                        except KeyError:
+                            return False
+                Default behaviour: remove all annotations
+            page: PageObject from which to remove annotations. Default: remove annotations from all pages.
+        """
+        if subtypes is not None:
+            if annotation_filter_function is not None:
+                logger_warning(
+                    "PdfWriter.remove_annotations: Argument annotation_filter_function was ignored"
+                    "because argument subtypes was set.",
+                    __name__,
+                )
+
+            def _annotation_filter_function(
+                page: DictionaryObject, an: ArrayObject, obj: DictionaryObject
+            ) -> bool:
+                return cast(str, obj["/Subtype"]) in cast(ArrayObject, subtypes)
+
+            annotation_filter_function = _annotation_filter_function
+
+        if annotation_filter_function is None:
+            annotation_filter_function = default_annotation_filter_function
+
         for page in self.pages:
-            self._remove_annots_from_page(page, subtypes)
+            self._remove_annots_from_page(page, annotation_filter_function)
 
     def _remove_annots_from_page(
         self,
         page: Union[IndirectObject, PageObject, DictionaryObject],
-        subtypes: Optional[Iterable[str]],
+        annotation_filter_function: Optional[Callable] = None,
     ) -> None:
+        """
+        Remove annotations using a custom delete_decide_function.
+
+        Args:
+            page: Page object to clean up.
+            annotation_filter_function: Function that takes three arguments,
+                DictionaryObject, ArrayObject and DictionaryObject, and decides whether to remove them. For example:
+                    def is_google_link(
+                        page: DictionaryObject,
+                        annotation: ArrayObject,
+                        obj: DictionaryObject
+                    ) -> bool:
+                        try:
+                            uri = obj['/A']['/URI']
+                            return uri.startswith('https://google.com/')
+                        except KeyError:
+                            return False
+                Default behaviour: remove all annotations
+        """
+        if annotation_filter_function is None:
+            annotation_filter_function = default_annotation_filter_function
+
         page = cast(DictionaryObject, page.get_object())
         if PG.ANNOTS in page:
             i = 0
             while i < len(cast(ArrayObject, page[PG.ANNOTS])):
                 an = cast(ArrayObject, page[PG.ANNOTS])[i]
```

Comment:
an normally contains the IndirectObject pointing onto obj. I do not see any use case where this is required. However if somebody needs it, it will be available through indirect_reference property.
For me, just the page, and the annotation (Just to be renamed in the example) which may be clearer are sufficient.

Reply:
Not sure I properly understood your comment. You're basically suggesting to have a simpler filter function?

```python
def default_annotation_filter_function(page: Any, an: Any, obj: Any) -> bool:
    becomes:
def default_annotation_filter_function(page: Any, an: Any) -> bool:
```

And `obj` would still be there in `an.indirect_reference`?

I never truly understood these objects. Can you maybe write a useful documentation for `annotation_filter_function` as [suggested here](https://github.com/py-pdf/pypdf/pull/1831#discussion_r1299423711)?